File: runtime/perl/units/rt/fnearest.pl
# requires: rt/quiet_nan
# C99 nearbyint under the default rounding mode is round-half-to-even and
# preserves signed zeros (measured, ADR-55) — exactly wasm's nearest.
use POSIX ();
sub fnearest {
    my $x = $_[0];
    return Rt::quiet_nan($x) if $x != $x;
    return POSIX::nearbyint($x);
}

File: crates/dewasm-backend-ruby/units/rt/fnearest.rb
# requires: rt/quiet_nan
def fnearest(x)
  return quiet_nan(x) if x.nan?
  return x unless x.finite?
  return x if x == 0.0
  r = x.round(half: :even)
  r == 0 && x < 0.0 ? -0.0 : r.to_f
end
